File: 131. Palindrome Partitioning.py
class Solution:
    def partition(self, s: str) -> List[List[str]]:
        result = []
        def palindrome(s):
            for i in range(0, len(s)//2):
                if s[i] != s[-i-1]: return False
            return True
        def populate(index, current: List[str]):
            nonlocal result, s
            if index >= len(s):
                result.append(current.copy())
                return
            for i in range(index, len(s)):
                sub = s[index:i+1]
                if palindrome(sub):
                    current.append(sub)
                    populate(i+1, current)
                    current.pop()
        populate(0, [])
        return result

File: 131. Palindrome Partitioning.go
func partition(s string) [][]string {
	result := [][]string{}
	palindrome := func(s string) bool {
		for i := 0; i < len(s)/2; i++ {
			if s[i] != s[len(s)-1-i] {
				return false
			}
		}
		return true
	}
	var populate func(int, []string)
	populate = func(index int, current []string) {
		if index >= len(s) {
			result = append(result, append([]string{}, current...))
			return
		}
		for i := index; i < len(s); i++ {
			sub := s[index : i+1]
			if palindrome(sub) {
				current = append(current, sub)
				populate(i+1, current)
				current = current[:len(current)-1]
			}
		}
	}
	populate(0, []string{})
	return result
}